snapyr/src/main/java/com/snapyr/sdk/notifications/SnapyrNotification.java

```diff
@@ -0,0 +1,227 @@
+            Log.e(
+                    "Snapyr",
+                    "onMessageReceived: Invalid message - encountered JSON error trying to parse payload JSON; returning.",
+                    e);
+            throw new IllegalArgumentException(
+                    "Invalid message - encountered JSON error trying to parse payload JSON");
+        }
+
+        this.titleText = getOrDefault(jsonData, SnapyrNotificationHandler.NOTIF_TITLE_KEY, null);
+        this.contentText =
+                getOrDefault(jsonData, SnapyrNotificationHandler.NOTIF_CONTENT_KEY, null);
+        if (titleText == null || contentText == null) {
+            throw new IllegalArgumentException("Invalid message - missing required data");
+        }
+        this.subtitleText =
+                getOrDefault(jsonData, SnapyrNotificationHandler.NOTIF_SUBTITLE_KEY, null);
+
+        String templateId;
+        Date templateModified;
+        try {
+            JSONObject template =
+                    jsonData.getJSONObject(SnapyrNotificationHandler.NOTIF_TEMPLATE_KEY);
+            templateId = template.getString("id");
+            String modifiedStr = jsonData.getString("modified");
+            templateModified = Utils.parseISO8601Date(modifiedStr);
+        } catch (Exception e) {
+            Log.e(
+                    "Snapyr",
+                    "Could not parse push template data. Continuing without template/action buttons.");
+            templateId = null;
+            templateModified = null;
+        }
+        this.templateId = templateId;
+        this.templateModified = templateModified;
+
+        this.channelId =
+                getOrDefault(
+                        jsonData,
+                        SnapyrNotificationHandler.NOTIF_CHANNEL_ID_KEY,
+                        SnapyrNotificationHandler.CHANNEL_DEFAULT_ID);
+        this.channelName =
+                getOrDefault(
+                        jsonData,
+                        SnapyrNotificationHandler.NOTIF_CHANNEL_NAME_KEY,
+                        SnapyrNotificationHandler.CHANNEL_DEFAULT_NAME);
+        this.channelDescription =
+                getOrDefault(
+                        jsonData,
+                        SnapyrNotificationHandler.NOTIF_CHANNEL_DESCRIPTION_KEY,
+                        SnapyrNotificationHandler.CHANNEL_DEFAULT_DESCRIPTION);
+
+        this.actionId =
+                getOrDefault(jsonData, SnapyrNotificationHandler.ACTION_ID_KEY, null); // unused?
+        this.actionToken = getOrDefault(jsonData, SnapyrNotificationHandler.NOTIF_TOKEN_KEY, null);
+
+        String deepLinkUrl =
+                getOrDefault(jsonData, SnapyrNotificationHandler.ACTION_DEEP_LINK_KEY, null);
+        this.deepLinkUrl = (deepLinkUrl != null) ? Uri.parse(deepLinkUrl) : null;
+
+        this.imageUrl = getOrDefault(jsonData, SnapyrNotificationHandler.NOTIF_IMAGE_URL_KEY, null);
+    }
+
+    protected SnapyrNotification(Parcel in) {
+        titleText = in.readString();
```

Comment:
maybe a comment here stating that the order matters and must match the parcel out?

Reply:
Good call, will add that